File: web/apps/control-center/src/domains/auth/mapper.test.ts
import { describe, it, expect } from 'vitest';
import { mapUser, mapUsers, mapRoles, mapPagedUsers } from './mapper';

describe('mapUser', () => {
  it('maps a full user response', () => {
    const user = mapUser({ id: '1', email: 'a@b.com', displayName: 'Alice', isDisabled: false, roles: ['admin'] });
    expect(user.id).toBe('1');
    expect(user.email).toBe('a@b.com');
    expect(user.displayName).toBe('Alice');
    expect(user.isDisabled).toBe(false);
    expect(user.roles).toEqual(['admin']);
  });

  it('falls back to name when displayName is missing', () => {
    const user = mapUser({ name: 'Bob' });
    expect(user.displayName).toBe('Bob');
  });

  it('uses fallbackId when id is missing', () => {
    const user = mapUser({}, 'fallback-id');
    expect(user.id).toBe('fallback-id');
  });

  it('defaults isDisabled to false', () => {
    const user = mapUser({});
    expect(user.isDisabled).toBe(false);
  });

  it('maps role objects by extracting name', () => {
    const user = mapUser({ roles: [{ name: 'editor', contextId: 'site-1' }] });
    expect(user.roles).toEqual(['editor']);
  });

  it('handles empty roles', () => {
    const user = mapUser({ roles: [] });
    expect(user.roles).toEqual([]);
  });
});

describe('mapUsers', () => {
  it('maps array of users', () => {
    const users = mapUsers([{ id: '1', email: 'a@b.com' }, { id: '2', email: 'c@d.com' }]);
    expect(users).toHaveLength(2);
    expect(users[0].email).toBe('a@b.com');
  });
});

describe('mapRoles', () => {
  it('maps role objects with contextId', () => {
    const roles = mapRoles({ roles: [{ name: 'admin', contextId: 'site-1' }] });
    expect(roles).toEqual([{ name: 'admin', contextId: 'site-1' }]);
  });

  it('maps role objects without contextId', () => {
    const roles = mapRoles({ roles: [{ name: 'viewer' }] });
    expect(roles).toEqual([{ name: 'viewer' }]);
  });

  it('filters out string roles', () => {
    const roles = mapRoles({ roles: ['string-role', { name: 'admin' }] });
    expect(roles).toEqual([{ name: 'admin' }]);
  });

  it('returns empty array when no roles', () => {
    const roles = mapRoles({});
    expect(roles).toEqual([]);
  });
});

describe('mapPagedUsers', () => {
  it('maps paged users response', () => {
    const result = mapPagedUsers({ data: [{ id: '1' }], total: 10, page: 1, pageSize: 5 });
    expect(result.total).toBe(10);
    expect(result.page).toBe(1);
    expect(result.pageSize).toBe(5);
    expect(result.data).toHaveLength(1);
  });
});

File: backend/__SLUG__/src/domains/auth/mapper_test.go
package auth

import (
	"testing"

	"go_backend/src/shared/pagination"
)

func newMapper() *Mapper {
	return NewMapper()
}

func TestUserToDTO(t *testing.T) {
	m := newMapper()
	user := &User{
		ID:    "u1",
		Email: "a@b.com",
		Name:  "Alice",
		Roles: []UserRole{{Name: "admin", ContextId: "site-1"}},
	}
	dto := m.UserToDTO(user)
	if dto.ID != "u1" {
		t.Errorf("expected ID u1, got %s", dto.ID)
	}
	if dto.Email != "a@b.com" {
		t.Errorf("expected email a@b.com, got %s", dto.Email)
	}
	if len(dto.Roles) != 1 || dto.Roles[0].Name != "admin" {
		t.Errorf("expected roles [admin], got %v", dto.Roles)
	}
}

func TestSessionToDTO(t *testing.T) {
	m := newMapper()
	session := &Session{
		UserID:        "u1",
		SessionHandle: "sh1",
		AccessToken:   "at1",
		RefreshToken:  "rt1",
	}
	dto := m.SessionToDTO(session)
	if dto.UserID != "u1" || dto.AccessToken != "at1" {
		t.Errorf("unexpected session dto: %+v", dto)
	}
}

func TestSignUpRequestToModel(t *testing.T) {
	m := newMapper()
	dto := &SignUpRequestDTO{Email: "x@y.com", Password: "pass123"}
	model := m.SignUpRequestToModel(dto)
	if model.Email != "x@y.com" {
		t.Errorf("expected email x@y.com, got %s", model.Email)
	}
}

func TestSignInRequestToModel(t *testing.T) {
	m := newMapper()
	dto := &SignInRequestDTO{Email: "x@y.com", Password: "pass"}
	model := m.SignInRequestToModel(dto)
	if model.Password != "pass" {
		t.Errorf("expected password pass, got %s", model.Password)
	}
}

func TestApplicationUserToDTO_NilRoles(t *testing.T) {
	m := newMapper()
	u := &ApplicationUser{ID: "u1", Email: "a@b.com", DisplayName: "Alice", IsDisabled: false, Roles: nil}
	dto := m.ApplicationUserToDTO(u)
	if len(dto.Roles) != 0 {
		t.Errorf("expected empty roles, got %v", dto.Roles)
	}
}

func TestApplicationUserToDTO_WithRoles(t *testing.T) {
	m := newMapper()
	u := &ApplicationUser{ID: "u2", Email: "b@c.com", DisplayName: "Bob", Roles: []string{"editor"}}
	dto := m.ApplicationUserToDTO(u)
	if len(dto.Roles) != 1 || dto.Roles[0] != "editor" {
		t.Errorf("expected [editor], got %v", dto.Roles)
	}
}

func TestApplicationUserListToDTO(t *testing.T) {
	m := newMapper()
	users := []ApplicationUser{
		{ID: "u1", Roles: []string{"admin"}},
		{ID: "u2", Roles: nil},
	}
	dtos := m.ApplicationUserListToDTO(users)
	if len(dtos) != 2 {
		t.Errorf("expected 2 dtos, got %d", len(dtos))
	}
}

func TestApplicationUserPageToDTO(t *testing.T) {
	m := newMapper()
	page := pagination.PagedResponse[ApplicationUser]{
		Data:     []ApplicationUser{{ID: "u1"}},
		Total:    1,
		Page:     0,
		PageSize: 10,
	}
	result := m.ApplicationUserPageToDTO(page)
	if result.Total != 1 || len(result.Data) != 1 {
		t.Errorf("unexpected paged result: %+v", result)
	}
}

func TestChangePasswordRequestToModel(t *testing.T) {
	m := newMapper()
	dto := &ChangePasswordRequestDTO{OldPassword: "old", NewPassword: "new"}
	model := m.ChangePasswordRequestToModel(dto)
	if model.OldPassword != "old" || model.NewPassword != "new" {
		t.Errorf("unexpected model: %+v", model)
	}
}
